### PR TITLE
Proper effect tracking for LVA

### PR DESCRIPTION
**Pull request by Anabra**
This PR is an implementation of the [effects.md](https://github.com/Anabra/grin/blob/master/notes/effects.md) and [liveness-and-effects.md](https://github.com/Anabra/grin/blob/master/notes/liveness-and-effects.md) proposals. Contrary to the latter proposal, the current implementation of LVA tracks effects for itself like the dedicated effect tracking analysis. However, it only tracks the existence of the effects (`Bool`) instead of tracking the actual effects (names of externals).

### CODE REVIEW

File: grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+import Grin.Grin
+import Grin.Pretty
+import Grin.TypeEnv
+import qualified AbstractInterpretation.IR as IR
+import AbstractInterpretation.IR (Instruction(..), AbstractProgram(..), AbstractMapping(..))
+import AbstractInterpretation.EffectTracking.CodeGenBase
+
+-- cannot return Z anywhere, any computation can contain side effects
+returnNewReg :: CG Result
+returnNewReg = do
```
Comment (andorp) [inline grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs]: Applicative style `R <$> newReg`

Comment (andorp) [reply grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs]: Couldn't be this a local function for the `codeGenM` if you don't intend to use anywhere else?

File: papers/haskell-exchange-2019/stg-grin/stg-grin.note [hunk 2]
```diff
@@ -0,0 +1,25 @@
+- STG in GHC pipeline (rewrite rules, core plugins, core opts before STG, core has types, STG has none)
```
Comment (andorp) [inline papers/haskell-exchange-2019/stg-grin/stg-grin.note]: I think this shouldn't be part of the PR. :)

File: papers/haskell-exchange-2019/stg-grin/stg-grin.abstract [hunk 3]
```diff
@@ -0,0 +1,5 @@
+A comparison between STG and GRIN
```
Comment (andorp) [inline papers/haskell-exchange-2019/stg-grin/stg-grin.abstract]: I think this shouldn't be part of the PR. :)

File: grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs (hunk 1 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs]: Done.

File: papers/haskell-exchange-2019/stg-grin/stg-grin.note (hunk 2 above)
Comment (Anabra) [reply papers/haskell-exchange-2019/stg-grin/stg-grin.note]: Done.

File: papers/haskell-exchange-2019/stg-grin/stg-grin.abstract (hunk 3 above)
Comment (Anabra) [reply papers/haskell-exchange-2019/stg-grin/stg-grin.abstract]: Done.

File: grin/app/GrinCLI.hs [hunk 4]
```diff
@@ -90,6 +90,11 @@ pipelineOpts =
   <|> flg (LVA PrintProgram) "print-lva-code" "Prints the live variable analysis machine"
   <|> flg (LVA RunPure) "run-lva-pure" "Runs the live variable analysis machine via pure interpreter"
   <|> flg (LVA PrintResult) "print-lva-result" "Prints the live variable analysis result"
+  <|> flg (ET Compile) "compile-et" "Compiles effect tracking analysis machine"
```
Comment (andorp) [inline grin/app/GrinCLI.hs]: This is getting out of hand :) In a later PR we should unify these steps, such as `--ai-compile=et`, `--ai-optimise=lva` etc.

No action required in this PR.

File: grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs [hunk 5]
```diff
@@ -109,11 +109,69 @@ codeGenVal = \case
...
+    emit IR.Store {srcReg = valReg, address = loc}
+    emit IR.Set {dstReg = r, constant = IR.CHeapLocation loc}
+    pure r
+
+codeGenExternal :: External -> [Val] -> CG Result
+codeGenExternal External{..} args = do
+  valRegs <- mapM codeGenVal args
+  argMap <- concat <$> zipWithM projectType valRegs eArgsType
+  r <- constructType argMap eRetType
+  pure $ R r
```
Comment (andorp) [inline grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs]: How about an applicative form?
```
R <$> constructType argMap eRetType
```

File: grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs [hunk 6]
```diff
@@ -109,11 +109,69 @@ codeGenVal = \case
...
+
 codeGen :: Exp -> (AbstractProgram, CByMapping)
 codeGen e = flip evalState emptyCGState $ para folder e >> mkCByProgramM where
   folder :: ExpF (Exp, CG Result) -> CG Result
   folder = \case
-    ProgramF exts defs -> (sequence_ . fmap snd $ defs) >> pure Z
+    -- ProgramF exts defs -> (sequence_ . fmap snd $ defs) >> pure Z
```
Comment (andorp) [inline grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs]: Please remove dead, commented out code.

File: grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs [hunk 7]
```diff
@@ -109,11 +109,69 @@ codeGenVal = \case
...
+    ProgramF exts defs -> do
+      mapM_ addExternal exts
+      sequence_ . fmap snd $ defs
```
Comment (andorp) [inline grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs]: Maybe `mapM_ snd defs` ?

File: grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+        emit IR.Move { srcReg = altReg, dstReg = caseResultReg }
+      pure $ R caseResultReg
+
+    AltF _ exp -> exp
+
+    SAppF name args -> getExternal name >>= \case
+      Just ext  -> do
+        appReg <- newReg
+        extID  <- fromJust <$> getExternalID name
+        if eEffectful ext then
```
Comment (andorp) [inline grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs]: Please use `when`

File: grin/src/AbstractInterpretation/EffectTracking/CodeGenBase.hs [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+import Grin.Grin (Name, SimpleType(..), CPat(..), unpackName, Tag(..), External(..))
+import Grin.TypeEnvDefs
+import AbstractInterpretation.IR (Instruction(..), Reg(..), AbstractMapping)
+import qualified AbstractInterpretation.IR as IR
+
+import Lens.Micro.Platform
+
+type ExternalID = Int32
+
+data External' = E { extID  :: ExternalID
```
Comment (andorp) [inline grin/src/AbstractInterpretation/EffectTracking/CodeGenBase.hs]: I am strongly against tick in names even if it for local data, because they are slippery slope. Please could you find a better name here?

Comment (andorp) [reply grin/src/AbstractInterpretation/EffectTracking/CodeGenBase.hs]: Also if you find one from me, hit me with a stick ;)

File: grin/src/AbstractInterpretation/EffectTracking/Result.hs [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+import Grin.Grin (Name, Tag, External(..))
+import Grin.Pretty
+import AbstractInterpretation.IR hiding (Tag, SimpleType)
+import AbstractInterpretation.Reduce (ComputerState(..))
+import qualified Grin.TypeEnv as TypeEnv
+import qualified AbstractInterpretation.IR as IR
+import qualified AbstractInterpretation.Reduce as R
+import AbstractInterpretation.EffectTracking.CodeGenBase (ETMapping(..))
+
```
Comment (andorp) [inline grin/src/AbstractInterpretation/EffectTracking/Result.hs]: Nit: I tend to leave two or three newlines between the last import and the first definition in a function. It is more pleasant to the eye.

File: grin/src/AbstractInterpretation/LiveVariable/CodeGen.hs [hunk 11]
```diff
@@ -196,53 +239,55 @@ codeGenM e = (cata folder >=> const setMainLive) e
...
+              Lit {} -> emit IR.Set { dstReg = lhsReg, constant = IR.CNodeItem irTag idx live }
+              _ -> error $ "illegal node pattern component " ++ show arg
+          emit IR.If
+            { condition     = IR.NodeTypeExists irTag
+            , srcReg        = lhsReg
+            , instructions  = bindInstructions
+            }
+        _ -> error $ "unsupported lpat " ++ show lpat
+
+      rhs <- rightExp
```
Comment (andorp) [inline grin/src/AbstractInterpretation/LiveVariable/CodeGen.hs]: What about `rhs@(R rhsRef) <- rightRxp` and `pure rhs` at the end?

File: grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs [hunk 12]
```diff
@@ -172,8 +184,11 @@ codeGenType cgSimpleTy cgNodeTy = \case
 isPointer :: IR.Predicate
 isPointer = IR.ValueIn (IR.Range 0 (maxBound :: Int32))
 
-isNotPointer :: IR.Predicate
-isNotPointer = IR.ValueIn (IR.Range (minBound :: Int32) 0)
+isLivenessInfo :: IR.Predicate
+isLivenessInfo = IR.ValueIn (IR.Range (-1) 0)
```
Comment (andorp) [inline grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs]: I suggest to code  `_SOME_NAME_ = (-1)`  to these constants. It improves the code readability and maintainability.

File: grin/src/AbstractInterpretation/Reduce.hs [hunk 13]
```diff
@@ -204,12 +205,20 @@ evalInstruction = \case
...
+      let mergedNthFields = mconcat . mapMaybe (V.!? n) . Map.elems $ tagMap
+      selectReg dstReg.simpleType %= (mappend mergedNthFields)
+
   Extend {..} -> do
     -- TODO: support all selectors
     value <- use $ selectReg srcReg.simpleType
     case dstSelector of
       NodeItem tag itemIndex -> selectTagMap dstReg.at tag.non mempty.ix itemIndex %= (mappend value)
       AllFields -> selectTagMap dstReg %= (Map.map (V.map (mappend value)))
+      -- TODO: review this
```
Comment (andorp) [inline grin/src/AbstractInterpretation/Reduce.hs]: Please could you comment on this? As a reader it is not clear how to and what to review?

File: grin/src/Grin/Grin.hs [hunk 14]
```diff
@@ -124,7 +125,7 @@ showTS :: Show a => a -> Name
 showTS = packName . show
 
 concatPrograms :: [Program] -> Program
-concatPrograms prgs = Program (concat exts) (concat defs) where
+concatPrograms prgs = Program (nub $ concat exts) (concat defs) where
```
Comment (andorp) [inline grin/src/Grin/Grin.hs]: I am not sure, if the `nub` is the right choice here. As we have the assumption that exts will be unique, if not we should throw an exception letting the clients know that they behave badly. 

What was the usecase when we created a duplicated ext joining programs?

File: grin/src/Pipeline/Pipeline.hs [hunk 15]
```diff
@@ -277,6 +282,8 @@ data PState = PState
     , _psCByResult      :: Maybe CBy.CByResult
     , _psLVAProgram     :: Maybe (IR.AbstractProgram, LVA.LVAMapping)
     , _psLVAResult      :: Maybe LVA.LVAResult
+    , _psETProgram      :: Maybe (IR.AbstractProgram, ET.ETMapping)
```
Comment (andorp) [inline grin/src/Pipeline/Pipeline.hs]: We need to improve the situation here, in a subsequent PR. See [Related card](https://github.com/grin-compiler/grin/projects/1#card-23109370 )

File: grin/grin/live-variable/case_fun_side_effect.grin [hunk 16]
```diff
@@ -0,0 +1,11 @@
+grinMain =
```
Comment (andorp) [inline grin/grin/live-variable/case_fun_side_effect.grin]: Please remove these example `.grin` files. We should put them into the repo when we have some end-to-end test case.

File: grin/app/GrinCLI.hs (hunk 4 above)
Comment (Anabra) [reply grin/app/GrinCLI.hs]: Yes, I agree.

File: grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs (hunk 5 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs]: Done.

File: grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs (hunk 6 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs]: Done.

File: grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs (hunk 7 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/CreatedBy/CodeGen.hs]: Done.

File: grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs (hunk 8 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/EffectTracking/CodeGen.hs]: Done.

File: grin/src/AbstractInterpretation/EffectTracking/CodeGenBase.hs (hunk 9 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/EffectTracking/CodeGenBase.hs]: Was thinking hard, but didn't find anything descriptive. Now I settled on `ExternalWithID`.

File: grin/src/AbstractInterpretation/EffectTracking/Result.hs (hunk 10 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/EffectTracking/Result.hs]: Done.

File: grin/src/AbstractInterpretation/LiveVariable/CodeGen.hs (hunk 11 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/LiveVariable/CodeGen.hs]: The problem is that there is no `MonadFail` instance for the code gen monad.

File: grin/src/AbstractInterpretation/EffectTracking/CodeGenBase.hs (hunk 9 above)
Comment (andorp) [reply grin/src/AbstractInterpretation/EffectTracking/CodeGenBase.hs]: This is way better than the tick :)

File: grin/src/AbstractInterpretation/LiveVariable/CodeGen.hs (hunk 11 above)
Comment (andorp) [reply grin/src/AbstractInterpretation/LiveVariable/CodeGen.hs]: Ah got it. I am not sure which is the better, to define a MonadFail for pattern match errors, or throw a runtime exception on pattern match error :) Leave it as it is.

File: grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs (hunk 12 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs]: In the `CodeGen` module, where they are used heavily, I did give them a name. Here, I didn't feel necessary, since the name of the function is quite descriptive. In my opinion, it would be more confusing to name these constants here, since we are talking about ranges. Something like `livenessLowerBound` would make much sense since that info is already readable from the definition. I'm open to other suggestions.

File: grin/src/AbstractInterpretation/Reduce.hs (hunk 13 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/Reduce.hs]: These are the new instructions added to the abstract machine. They are used in LVA at the default case. They set all the "tags" live. Tags are the very first fields of a node in this representation. I did not find any other instructions that could express this, that's why I added a new instruction.

File: grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs (hunk 12 above)
Comment (andorp) [reply grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs]: Ok, this is an implementation detail. But what is the interpretation of the (-1) in this particular case?

File: grin/src/Grin/Grin.hs (hunk 14 above)
Comment (Anabra) [reply grin/src/Grin/Grin.hs]: If the program already contains external definitions and we also add the primops prelude "manually" through the pipeline, the externals get duplicated. That's why I introduced this extra `nub` here.

File: grin/src/Pipeline/Pipeline.hs (hunk 15 above)
Comment (Anabra) [reply grin/src/Pipeline/Pipeline.hs]: Agreed.

File: grin/src/AbstractInterpretation/Reduce.hs (hunk 13 above)
Comment (andorp) [reply grin/src/AbstractInterpretation/Reduce.hs]: Got it, thanks.

File: grin/grin/live-variable/case_fun_side_effect.grin (hunk 16 above)
Comment (Anabra) [reply grin/grin/live-variable/case_fun_side_effect.grin]: Done.

**Comment by Anabra** [comment]
The input was great. Thanks for the feedback!

File: grin/src/Grin/Grin.hs (hunk 14 above)
Comment (andorp) [reply grin/src/Grin/Grin.hs]: @csabahruska What do you think?

File: grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs (hunk 12 above)
Comment (Anabra) [reply grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs]: `(-1)` is the representation of abstract value "live". `(-2)` is the same for "has effect".
I will think of a name. Do you have any suggestions?

File: grin/src/AbstractInterpretation/Reduce.hs [hunk 17]
```diff
@@ -204,12 +205,18 @@ evalInstruction = \case
       let mergedFields = mconcat . (map Data.Foldable.fold) . Map.elems $ tagMap
       selectReg dstReg.simpleType %= (mappend mergedFields)
 
+    EveryNthField n -> do
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/AbstractInterpretation/Reduce.hs]: I described it above:

> These are the new instructions added to the abstract machine. They are used in LVA at the default case. They set all the "tags" live. Tags are the very first fields of a node in this representation. I did not find any other instructions that could express this, that's why I added a new instruction.

File: grin/src/Transformations/Optimising/CSE.hs [hunk 18]
```diff
@@ -19,7 +19,7 @@ type Env = (Map SimpleExp SimpleExp)
 -- TODO: track if function parameters with location type can be updated in the called function to improve CSE
 
 commonSubExpressionElimination :: TypeEnv -> EffectMap -> Exp -> Exp
-commonSubExpressionElimination typeEnv effMap e = hylo skipUnit builder (mempty, e) where
+commonSubExpressionElimination typeEnv effectMap e = hylo skipUnit builder (mempty, e) where
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/Transformations/Optimising/CSE.hs]: Reverted the change.

File: grin/src/Transformations/Optimising/DeadVariableElimination.hs [hunk 19]
```diff
@@ -28,10 +28,10 @@ import Control.Monad.Trans.Except
 import Grin.Grin
 import Grin.Pretty
 import Grin.TypeEnv
-import Grin.EffectMap
 import Transformations.Util
-import AbstractInterpretation.LiveVariable.Result as LVA
+import AbstractInterpretation.LiveVariable.Result   as LVA
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/src/Transformations/Optimising/DeadVariableElimination.hs]: Done.

File: grin/test/Transformations/Optimising/CSESpec.hs [hunk 20]
```diff
@@ -15,10 +18,16 @@ import Transformations.EffectMap
 runTests :: IO ()
 runTests = hspec spec
 
-cseOptNoEff (tyEnv, exp) = commonSubExpressionElimination tyEnv mempty exp
+pipelineSrcWithCtx :: ((TypeEnv, Exp) -> (TypeEnv, Exp)) ->
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/test/Transformations/Optimising/CSESpec.hs]: Alright, will make a separate PR.

File: grin/test/Transformations/Optimising/DeadVariableEliminationSpec.hs [hunk 21]
```diff
@@ -39,9 +41,10 @@ spec = do
     it "heap" $ do
       let before = [prog|
             grinMain =
-              p0 <- store (CInt 0)
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/test/Transformations/Optimising/DeadVariableEliminationSpec.hs]: Some of them are mistakes, sorry about those. The rest is about introducing the new syntax conventions. LVA already relies on them in some cases. To make the test cases more homogenous, I introduced the new conventions everywhere.

File: grin/test/Transformations/Optimising/DeadVariableEliminationSpec.hs [hunk 22]
```diff
@@ -57,7 +60,8 @@ spec = do
       let before = [prog|
             grinMain =
               p0 <- store (CInt 0)
-              update p0 (CInt 1)
```
[the remark this answers is not in the document]
Comment (Anabra) [reply grin/test/Transformations/Optimising/DeadVariableEliminationSpec.hs]: Some of them are mistakes, sorry about those. The rest is about introducing the new syntax conventions. LVA already relies on them in some cases. To make the test cases more homogenous, I introduced the new conventions everywhere.

File: grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs (hunk 12 above)
Comment (andorp) [reply grin/src/AbstractInterpretation/LiveVariable/CodeGenBase.hs]: It is dead simple: _LIVE_ and _HAS_EFFECT_ .

File: grin/src/Grin/Grin.hs (hunk 14 above)
Comment (andorp) [reply grin/src/Grin/Grin.hs]: Than I am also ok with the `nub`

**Comment by andorp** [comment]
LGTM

File: grin/test/AbstractInterpretation/EffectTrackingSpec.hs [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+spec = describe "Effect Tracking Analysis" $ do
+
+  it "simple_print" $ do
+    let exp = [prog|
+          grinMain =
+            _prim_int_print 0
+        |]
+    let expected = mempty
+          { _register = []
+          , _function = [ ("grinMain", Effects ["_prim_int_print"]) ]
```
[the remark this answers is not in the document]
Comment (csabahruska) [reply grin/test/AbstractInterpretation/EffectTrackingSpec.hs]: Did you add tests for externals?

[the remark this answers is not in the document]
Comment (Anabra) [reply grin/test/AbstractInterpretation/EffectTrackingSpec.hs]: The effect tracking analysis calculates only the names of the effects. It is based on externals, not primops. Are you suggesting we should collect the `External`s instead of just their names during the analysis?

[the remark this answers is not in the document]
Comment (csabahruska) [reply grin/test/AbstractInterpretation/EffectTrackingSpec.hs]: Do not rely on predefined externals in that test. Define you own external function/primop.

[the remark this answers is not in the document]
Comment (Anabra) [reply grin/test/AbstractInterpretation/EffectTrackingSpec.hs]: Alright, i added some tests. Also added an error message to the effect tracking analysis in case the program tried to redefine an already existing external.

**Comment by Anabra** [comment]
> if `newPrimSomething#` is pure then why is it marked effectful?

The # version is always effectful, isn't?
